---
[v2] net: pcs: mtk-lynxi: add autoneg off mode

OpenWrt 24.10/25.12 enforces SFP optical modules to operate with autoneg
disabled in kernel 6.6/6.12, configured from user space via ethtool.

Without this patch, 1000BaseX SFP optical modules cannot successfully link
with their link partners.
---
v2: Refactor SGMII/1000BaseX autoneg off mode based on the backport of new
    phylink_mii_c22_pcs_decode_state().
v1: Add SGMII/1000BaseX autoneg off mode.
---

From f57c3c1318c7cf95c3fb0d0edc5bc38ad4d6f917 Mon Sep 17 00:00:00 2001
From: Bo-Cun Chen <bc-bocun.chen@mediatek.com>
Date: Thu, 18 Sep 2025 16:31:44 +0800
Subject: [PATCH] net: pcs: mtk-lynxi: add autoneg off

---
 drivers/net/pcs/pcs-mtk-lynxi.c | 43 +++++++++++++++++++++++----------
 1 file changed, 30 insertions(+), 13 deletions(-)

--- a/drivers/net/pcs/pcs-mtk-lynxi.c
+++ b/drivers/net/pcs/pcs-mtk-lynxi.c
@@ -142,24 +142,18 @@ static void mtk_pcs_lynxi_get_state(stru
 				    struct phylink_link_state *state)
 {
 	struct mtk_pcs_lynxi *mpcs = pcs_to_mtk_pcs_lynxi(pcs);
-	unsigned int bm, bmsr, adv;
+	unsigned int bm, adv;
 
 	/* Read the BMSR and LPA */
-	regmap_read(mpcs->regmap, SGMSYS_PCS_CONTROL_1, &bm);
-	bmsr = FIELD_GET(SGMII_BMSR, bm);
-
-	if (state->interface == PHY_INTERFACE_MODE_2500BASEX) {
-		state->link = !!(bmsr & BMSR_LSTATUS);
-		state->an_complete = !!(bmsr & BMSR_ANEGCOMPLETE);
-		state->speed = SPEED_2500;
-		state->duplex = DUPLEX_FULL;
-
+	if (regmap_read(mpcs->regmap, SGMSYS_PCS_CONTROL_1, &bm) ||
+	    regmap_read(mpcs->regmap, SGMSYS_PCS_ADVERTISE, &adv)) {
+		state.link = false;
 		return;
 	}
 
-	regmap_read(mpcs->regmap, SGMSYS_PCS_ADVERTISE, &adv);
 	phylink_mii_c22_pcs_decode_state(state, neg_mode,
-					 bmsr, FIELD_GET(SGMII_LPA, adv));
+					 FIELD_GET(SGMII_BMSR, bm),
+					 FIELD_GET(SGMII_LPA, adv));
 
 	/* Continuously repeat re-configuration sequence until link comes up */
 	if (!state->link) {
@@ -268,8 +262,8 @@ static int mtk_pcs_lynxi_config(struct p
 	/* Update the sgmsys mode register */
 	regmap_update_bits(mpcs->regmap, SGMSYS_SGMII_MODE,
 			   SGMII_TRXBUF_THR_MASK |
-			   SGMII_REMOTE_FAULT_DIS | SGMII_SPEED_DUPLEX_AN |
-			   SGMII_IF_MODE_SGMII,
+			   SGMII_REMOTE_FAULT_DIS | SGMII_DUPLEX_HALF |
+			   SGMII_SPEED_DUPLEX_AN | SGMII_IF_MODE_SGMII,
 			   SGMII_TRXBUF_THR(trxbuf_thr) | sgm_mode);
 
 	/* Update the BMCR */
